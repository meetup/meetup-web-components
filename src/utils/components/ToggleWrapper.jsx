import PropTypes from 'prop-types';
import React from 'react';
import ConditionalWrap from './ConditionalWrap';

/**
 * Provides boolean isActive prop to wrapped component.
 * Prop name can also be customized if need be
 *
 * @param {React.element} ToggleComponent - the component to wrap
 * @returns {React.element}
 */
// export default function withToggleControl(WrappedComponent) {
/**
 * @module ToggleWrapper
 */
class ToggleWrapper extends React.Component {
	constructor(props) {
		super(props);

		this.state = { isActive: props.isActive };
		this.toggleActive = this.toggleBool.bind(this);
		this.onKeyUp = this.onKeyUp.bind(this);
	}

	componentWillReceiveProps(nextProps) {
		if (nextProps.isActive !== this.props.isActive) {
			this.toggleActive();
		}
	}

	toggleBool(e) {
		this.setState({ isActive: !this.state.isActive });

		if (this.props.onToggle) {
			this.props.onToggle(e);
		}
	}

	onKeyUp(e) {
		const isActivatingButton = [
			' ', /* space bar */
			'Enter'
		].some(key => e.key === key);

		if (isActivatingButton) {
			this.toggleActive();
		}
	}

	render() {
		const isInput = this.props.type == 'radio' || this.props.type == 'checkbox';

		return (
			<ConditionalWrap
				condition={!isInput}
				wrap={children => (
					<div
						role="button"
						aria-pressed={this.state.isActive}
						onKeyUp={this.onKeyUp}
						tabIndex={this.props.tabIndex || "0"}
					>
						{this.props.children({
							isActive: this.state.isActive,
							toggleActive: this.toggleActive,
						})}
					</div>
				)}
			>
				{this.props.children({
					tabIndex: isInput && this.props.tabIndex || "0",
					isActive: this.state.isActive,
					toggleActive: this.toggleActive,
					onKeyUp: this.props.type !== 'radio' && this.onKeyUp
				})}
			</ConditionalWrap>
		);
	}
}

ToggleWrapper.defaultProps = {
	isActive: false
};
ToggleWrapper.propTypes = {
	isActive: PropTypes.bool
};

export default ToggleWrapper;
