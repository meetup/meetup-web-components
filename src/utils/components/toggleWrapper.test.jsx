import React from 'react';
import { mount } from 'enzyme';

import ToggleWrapper from './ToggleWrapper';
import ConditionalWrap from './ConditionalWrap';

const PLAIN_COMPONENT_CLASS = 'plainComponent';

const onToggle = jest.fn();

/**
 * @module PlainComponent
 */
class PlainComponent extends React.PureComponent {
	render() {

		return(
			<ToggleWrapper
				onToggle={onToggle}
				tabIndex={this.props.tabIndex}
			>
				{({
					tabIndex,
					isActive,
					toggleActive,
					onKeyUp,
				}) => (
					<div
						aria-pressed={isActive}
						onClick={toggleActive}
						className={PLAIN_COMPONENT_CLASS}
						{...this.props}
					/>
				)}
			</ToggleWrapper>
		);
	}
}

/**
 * @module InputComponent
 */
class InputComponent extends React.PureComponent {
	render() {
		return(
			<ToggleWrapper
				type={this.props.inputType}
				isActive={this.props.isActive}
			>
				{({
					tabIndex,
					isActive,
					toggleActive,
					onKeyUp,
				}) => (
					<input
						type={this.props.inputType}
						checked={isActive}
						onChange={toggleActive}
						onKeyUp={onKeyUp}
						tabIndex={tabIndex}
						value="lol nothing matters"
						{...this.props}
					/>
				)}
			</ToggleWrapper>
		);
	}
}

describe('withLoading', function() {
	it('calls onToggle if it is passed', () => {
		const plainComponent = mount(<PlainComponent />);
		const plainComponentNode = plainComponent.find(`.${PLAIN_COMPONENT_CLASS}`);

		expect(onToggle).not.toHaveBeenCalled();
		plainComponentNode.simulate('click');
		expect(onToggle).toHaveBeenCalled();
	});

	describe('when this.props.type is not checkbox or radio', () => {
		let plainComponent, plainComponentNode;

		beforeEach(() => {
			plainComponent = mount(<PlainComponent />);
			plainComponentNode = plainComponent.find(`.${PLAIN_COMPONENT_CLASS}`);
		});
		afterEach(() => {
			plainComponent = null;
			plainComponentNode = null;
		});

		it('matches snapshot', () => {
			expect(plainComponent).toMatchSnapshot();
		});

		it('wraps children in an accessible wrapper component', () => {
			const wrapperComponent = plainComponent.find(ConditionalWrap).find('[role="button"]');

			expect(wrapperComponent.length).toBe(1);
		});

		it('passes props wrapper component', () => {
			const plainComponent = mount(<PlainComponent tabIndex="1" />);
			const wrapperComponent = plainComponent.find(ConditionalWrap).find('[role="button"]');

			expect(wrapperComponent.prop('tabIndex')).toBe("1");
		});

		it('isActive state updates using child component handler', () => {
			expect(plainComponent.find(ToggleWrapper).instance().state.isActive).toBe(false);
			plainComponentNode.simulate('click');
			expect(plainComponent.find(ToggleWrapper).instance().state.isActive).toBe(true);
		});

		it('isActive state updates on valid key press (Enter or Space bar)', () => {
			expect(plainComponent.find(ToggleWrapper).instance().state.isActive).toBe(false);
			plainComponentNode.simulate('keyUp', {key: 'Enter'});
			expect(plainComponent.find(ToggleWrapper).instance().state.isActive).toBe(true);
		});
	});

	describe('when this.props.type is checkbox', () => {
		let checkboxComponent;

		beforeEach(() => {
			checkboxComponent = mount(<InputComponent inputType="checkbox" />);
		});
		afterEach(() => {
			checkboxComponent = null;
		});

		it('matches snapshot', () => {
			expect(checkboxComponent).toMatchSnapshot();
		});

		it('isActive state updates onChange', () => {
			expect(checkboxComponent.find(ToggleWrapper).instance().state.isActive).toBe(false);
			checkboxComponent.find('input').simulate('change');
			expect(checkboxComponent.find(ToggleWrapper).instance().state.isActive).toBe(true);
		});

	});

	describe('when this.props.type is radio', () => {
		let radioComponent;

		beforeEach(() => {
			radioComponent = mount(<InputComponent inputType="radio" />);
		});
		afterEach(() => {
			radioComponent = null;
		});

		it('matches snapshot', () => {
			expect(radioComponent).toMatchSnapshot();
		});

		it('does not pass internal onKeyUp to children', () => {
			const checkboxKeyUp = radioComponent.find('input').prop('onKeyUp');
			const toggleWrapperKeyUp = radioComponent.find(ToggleWrapper).instance().onKeyUp;

			expect(checkboxKeyUp == toggleWrapperKeyUp).toBe(false);
		});
	});
});
