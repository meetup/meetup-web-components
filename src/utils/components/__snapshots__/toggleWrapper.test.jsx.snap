// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLoading when this.props.type is checkbox matches snapshot 1`] = `
<InputComponent
  inputType="checkbox"
>
  <ToggleWrapper
    isActive={false}
    type="checkbox"
  >
    <ConditionalWrap
      condition={false}
      wrap={[Function]}
    >
      <input
        checked={false}
        inputType="checkbox"
        onChange={[Function]}
        onKeyUp={[Function]}
        tabIndex="0"
        type="checkbox"
        value="lol nothing matters"
      />
    </ConditionalWrap>
  </ToggleWrapper>
</InputComponent>
`;

exports[`withLoading when this.props.type is not checkbox or radio matches snapshot 1`] = `
<PlainComponent>
  <ToggleWrapper
    isActive={false}
    onToggle={[Function]}
    tabIndex={undefined}
  >
    <ConditionalWrap
      condition={true}
      wrap={[Function]}
    >
      <div
        aria-pressed={false}
        onKeyUp={[Function]}
        role="button"
        tabIndex="0"
      >
        <div
          aria-pressed={false}
          className="plainComponent"
          onClick={[Function]}
        />
      </div>
    </ConditionalWrap>
  </ToggleWrapper>
</PlainComponent>
`;

exports[`withLoading when this.props.type is radio matches snapshot 1`] = `
<InputComponent
  inputType="radio"
>
  <ToggleWrapper
    isActive={false}
    type="radio"
  >
    <ConditionalWrap
      condition={false}
      wrap={[Function]}
    >
      <input
        checked={false}
        inputType="radio"
        onChange={[Function]}
        onKeyUp={false}
        tabIndex="0"
        type="radio"
        value="lol nothing matters"
      />
    </ConditionalWrap>
  </ToggleWrapper>
</InputComponent>
`;
